envoyproxy/envoy
# thrift filter: support skip decoding data after metadata in the thrift message

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/master/PULL_REQUESTS.md)

Commit Message:

Fixes https://github.com/envoyproxy/envoy/issues/13082 

Support skip decoding data after metadata in the thrift message. 

API change:

```
message ThriftProxy {
  option (udpa.annotations.versioning).previous_message_type =
      "envoy.config.filter.network.thrift_proxy.v2alpha1.ThriftProxy";

  ...

  // If set to true, Envoy will try to skip decode data after metadata in the Thrift message.
  // This mode will only work if the upstream and downstream protocols are the same and the transport
  // is the same, the transport type is framed and the protocol is not Twitter. Otherwise Envoy will
  // fallback to decode the data.
  bool payload_passthrough = 6;
}
```

Additional Description:

In payload_passthrough mode, there are some issues:

- Envoy cannot detect some errors and exceptions ( e.g. a reply that contains exceptions ). It's possible to improve this by peeking beginning of the payload.

- `payload_passthrough` controls both request and response path. It can be split into two options if we want more fine-grained control.

- `FilterStatus passthroughData(Buffer::Instance& data, uint64_t bytes_to_passthrough)` will not prohibit custom filters to modify buffer. Now it is assumed custom filters won't do that, otherwise behavior is undefined.

TODO: 

-  [x] add unit tests to conn_manager

Risk Level: Medium
Testing:

- unit test:
    - config
    - decoder
    - router
    - conn_manager

- integration:
   - add an parameter  `payload_passthrough` 

- manual:
   - send requests and verify responses 

Docs Changes:

TODO

Release Notes:

TODO

Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
